app/src/main/java/org/schabi/newpipe/player/mediabrowser/MediaBrowserConnector.kt

```diff
@@ -0,0 +1,738 @@
+                    )
+                )
+                mediaItems.add(
+                    createRootMediaItem(
+                        ID_HISTORY,
+                        playerService.resources.getString(R.string.action_history),
+                        R.drawable.ic_history_white
+                    )
+                )
+                return Single.just(mediaItems)
+            }
+
+            val uriType = path[0]
+            path.removeAt(0)
+
+            when (uriType) {
+                ID_BOOKMARKS -> {
+                    if (path.isEmpty()) {
+                        return populateBookmarks()
+                    }
+                    if (path.size == 2) {
+                        val localOrRemote = path[0]
+                        val playlistId = path[1].toLong()
+                        if (localOrRemote == ID_LOCAL) {
+                            return populateLocalPlaylist(playlistId)
+                        } else if (localOrRemote == ID_REMOTE) {
+                            return populateRemotePlaylist(playlistId)
+                        }
+                    }
+                    Log.w(TAG, "Unknown playlist URI: $parentId")
+                    throw parseError()
+                }
+
+                ID_HISTORY -> return populateHistory()
+                else -> throw parseError()
+            }
+        } catch (e: ContentNotAvailableException) {
+            return Single.error(e)
+        }
+    }
+
+    private fun populateHistory(): Single<List<MediaBrowserCompat.MediaItem>> =
+        database
+            .streamHistoryDAO()
+            .history
+            .firstOrError()
+            .map { items -> items.map(::createHistoryMediaItem) }
+
+    private fun createHistoryMediaItem(streamHistoryEntry: StreamHistoryEntry): MediaBrowserCompat.MediaItem {
+        val builder = MediaDescriptionCompat.Builder()
+        val mediaId = buildMediaId()
+            .appendPath(ID_HISTORY)
+            .appendPath(streamHistoryEntry.streamId.toString())
+            .build().toString()
+        builder.setMediaId(mediaId)
+            .setTitle(streamHistoryEntry.streamEntity.title)
+            .setSubtitle(streamHistoryEntry.streamEntity.uploader)
+            .setIconUri(Uri.parse(streamHistoryEntry.streamEntity.thumbnailUrl))
+
+        return MediaBrowserCompat.MediaItem(
+            builder.build(),
+            MediaBrowserCompat.MediaItem.FLAG_PLAYABLE
+        )
+    }
+
+    private fun setupBookmarksNotifications() {
+        bookmarksNotificationsDisposable =
+            playlists.subscribe { _ ->
+                playerService.notifyChildrenChanged(ID_BOOKMARKS)
+            }
+    }
+
+    private fun disposeBookmarksNotifications() {
+        bookmarksNotificationsDisposable?.dispose()
+    }
+
+    // Suppress Sonar warning replace list collection by Stream.toList call, as this method is only
+    // available in Android API 34 and not currently available with desugaring
```

Comment:
This was in the Java version of the class and should be removed.

Reply:
Sure, I'll remove it.